src/Tools/CompilerBenchmarks/EmitBenchmark.cs

```diff
@@ -0,0 +1,115 @@
+
+using System.IO;
+using BenchmarkDotNet.Attributes;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Emit;
+using static Microsoft.CodeAnalysis.Compilation;
+
+namespace CompilerBenchmarks
+{
+    public class EmitBenchmark
+    {
+        public enum EmitStageSelection
+        {
+            FullEmit,
+            SerializeOnly
+        }
+
+        [Params(EmitStageSelection.FullEmit, EmitStageSelection.SerializeOnly)]
+        public EmitStageSelection Selection;
+
+        private Compilation _comp;
+        private CommonPEModuleBuilder _moduleBeingBuilt;
+        private EmitOptions _options;
+
+        [GlobalSetup]
+        public void Setup()
+        {
+            _comp = Helpers.CreateReproCompilation();
+
+            // Call GetDiagnostics to force binding to finish and most semantic analysis to be completed
+            _ = _comp.GetDiagnostics();
+
+            if (Selection == EmitStageSelection.SerializeOnly)
+            {
+                _options = EmitOptions.Default.WithIncludePrivateMembers(true);
+
+                bool embedPdb = _options.DebugInformationFormat == DebugInformationFormat.Embedded;
+
+                var diagnostics = DiagnosticBag.GetInstance();
+
+                _moduleBeingBuilt = _comp.CheckOptionsAndCreateModuleBuilder(
+                    diagnostics,
+                    manifestResources: null,
+                    _options,
+                    debugEntryPoint: null,
+                    sourceLinkStream: null,
+                    embeddedTexts: null,
+                    testData: null,
+                    cancellationToken: default);
+
+                bool success = false;
+
+                success = _comp.CompileMethods(
+                    _moduleBeingBuilt,
+                    emittingPdb: embedPdb,
+                    emitMetadataOnly: _options.EmitMetadataOnly,
+                    emitTestCoverageData: _options.EmitTestCoverageData,
+                    diagnostics: diagnostics,
+                    filterOpt: null,
+                    cancellationToken: default);
+
+                _comp.GenerateResourcesAndDocumentationComments(
+                    _moduleBeingBuilt,
+                    xmlDocumentationStream: null,
+                    win32ResourcesStream: null,
+                    _options.OutputNameOverride,
+                    diagnostics,
+                    cancellationToken: default);
+
+                _comp.ReportUnusedImports(null, diagnostics, default);
+                _moduleBeingBuilt.CompilationFinished();
+
+                diagnostics.Free();
+            }
+        }
+
+        [Benchmark]
+        public object RunEmit()
+        {
+            switch (Selection)
+            {
+                case EmitStageSelection.FullEmit:
+                {
+                    var stream = new MemoryStream();
```

Comment:
> var stream = new MemoryStream(); [](start = 19, length = 33)

The default ctor for MemoryStream will allocate `byte[]` under the hood. Those allocations will end up getting included in the BenchMmark data. Should we pre-allocate a larger buffer here to keep the noise down?